Change representation of class signatures

lpw25 opened the pull request:
This PR refactors code that handles class signatures and structures. It fixes a number of bugs including #7894 and improves error messages.

The main change is to replace the current representation of signatures:
```ocaml
type class_signature =
  { csig_self: type_expr;
    csig_vars:
      (Asttypes.mutable_flag * Asttypes.virtual_flag * type_expr) Vars.t;
    csig_concr: Concr.t;
    csig_inher: (Path.t * type_expr list) list }
```
 where `csig_self` is restricted to be exactly a `Tobject` type and methods are accessed by fetching them out of this `Tobject` node, with:
```ocaml
  { csig_self: type_expr;
    mutable csig_self_row: type_expr;
    mutable csig_vars: (mutable_flag * virtual_flag * type_expr) Vars.t;
    mutable csig_meths: (private_flag * virtual_flag * type_expr) Meths.t }
```
where methods are represented directly in a map. The fields are also made mutable to allow `class_signature` to be treated more like `type_expr` and mutated as it is inferred. The restrictions on `csig_self`, which have been the cause of a number of bugs in the past, are removed.

Some of the changes are not so much bug fixes as slight changes to the design. These are mostly related to instance variables, whose behaviour was changed from shadowing to overriding in OCaml 3.10, where some remnants of the old shadowing behaviour seem to have persisted.

The first commit introduces a lot of new tests around classes. Each commit after that which is expected to change observable behaviour should have at least one test whose behaviour changes.

lpw25 committed:
Make the two-passes of class_field explicit

lpw25 committed:
Collect vars and meths without references

lpw25 committed:
Add more classes tests

lpw25 committed:
Relax object duplication restrictions

Currently object duplications (i.e. `{< ... >}`) can only refer to
instance variables that are specified before the method in which
they occur. This restriction made sense before OCaml 3.10 when
instance variables could shadow one another. However, instance
variables now override one another instead and so the restriction
is not needed.

lpw25 committed:
Represent ancestor variables more directly

Previously, some of the translation of method calls on ancestor
variables (bindings from `inherit` statements) was done in typecore.ml
by creating fake typedtree expressions. This has been moved into
translcore.ml which is simpler. We also pass the types of inherited
methods directly as part of the ancestor variable's kind, which
eleminates the need for the dummy `selfpat-n` variables.

lpw25 committed:
Track dummy methods via their scope

Prevent self types from escaping using the scope of the dummy
method field rather than its level. This is a more accurate
representation of the restriction and allows the level of the
self type to vary, which fixes a bug in principal mode.

lpw25 committed:
Remove private_self and public_self

Now that private methods are stored in the method table there is
no need to distinguish `private_self`, `public_self` and `self_type`
in `Typeclass.class_structure`. This required extending
`Ctype.filter_method` to handle the case of filtering a private method
from a closed object type -- which brings the function into line with
the equivalent call to `Ctype.unify`.

lpw25 committed:
Fix warning attributes in class signatures

[@@@warning ...] was not previously preserved in class signatures

lpw25 committed:
Set object name of self type

Sets the object name of the self type to the #-abbreviation, which
improves error messages and allows us to remove the `unify_parents`
functions from Typeclass.

lpw25 committed:
Keep class signature row up-to-date

Keep the csig_self_row field up-to-date as we add new methods. This
seems more correct and should be slightly more efficient in the
common case.

lpw25 committed:
Rename Concr

The Types.Concr module is just a string set, and it is used in
a number of places. It was originally used for the set of
concrete methods in a class signature, but it's not used for
that anymore and almost none of the places it is still used
are related to that. This commit replaces is with MethSet and
VarSet modules for sets of method or instance variable labels,
and with Misc.Stdlib.String.Set for other uses.

lpw25 committed:
Change warning 13 message

Remove the mention of OCaml 3.10 in the message of warning 13 (instance
variable override). It's been over a decade since that release -- I
think we can stop telling people that it changed the behaviour of
instance variables.

lpw25 committed:
Move method spine generalisation logic into Ctype

Move the method spine generalisation code into ctype.ml. Since that
is the only code that needs to use generalize_spine we also hide
generalize_spine from the interface of Ctype and rewrite the other
use to use generalize_structure. (I would prefer to get rid of
generalize_spine and use generalize_structure for methods as well,
since generalize_spine seems a bit suspect in terms of principality,
but that change would be very invovlved).

lpw25 committed:
Add dummy methods in all cases

Fix some regressions introduced 4.07 by ensuring that class types and
class expressions for classes under construction always include a dummy
method.

lpw25 committed:
Give more precise errors for virtual methods

Previously, the check that a non-virtual class did not
have virtual methods was delayed until the last stage of
type-checking the class. Now we do those checks more eagerly
allowing for clearer errors, especially in common cases.

lpw25 committed:
Remove unused csig_inher field

garrigue commented:
Could you first explain which bugs you are fixing (other than #7894), and where they come from.
You seem to be introducing some extensive changes, with a high risk of introducing new bugs, so I would first like to understand why this is needed (rather than just rolling back the removal of the dummy method).
The idea of making the information in class signatures more easily usable is not bad, but the output of the test suite suggests that this had at least some impact of the sharing of types, which may matter.

lpw25 commented:
> Could you first explain which bugs you are fixing (other than #7894), and where they come from.

I don't think most of them have bug reports. The easiest way to see them is to read through the commits one at a time. The changes to the tests in each commit should make clear what the bug is. I've even written some explanations in the commit messages for each of the changes.

> You seem to be introducing some extensive changes, with a high risk of introducing new bugs, so I would first like to understand why this is needed

The main intention is really to make this code simpler and easier to maintain. The existing code is very intricate and easy to break. That means that this patch indeed has a high risk of introducing bugs, but that is an incentive to merge the patch rather than not to, since it should remove this risk for any later patch that touches this code.

alainfrisch commented on typing/typeclass.ml:
```diff
@@ -259,54 +259,56 @@ let make_method loc cl_num expr =
-      inherit_class_type ~strict:false loc env sign parent.cltyp_type;
-      mkctf (Tctf_inherit parent)
-
+      Builtin_attributes.warning_scope ctf.pctf_attributes
```
It's a bit unfortunate (albeit very minor) that we need to duplicate the "logic" to handle pctf_attributes for all cases (except Pctf_attribute).  One could have another helper function that encapsule the use of `Builtin_attributes.warning_scope ctf.pctf_attributes` and `mkctf`:

````ocaml
let mkctf_attr f = Builtin_attributes.warning_scope ctf.pctf_attributes (mkctf (f ())) in
````

lpw25 replied:
That does seem slightly better.

lpw25 replied:
Fixed

alainfrisch commented on typing/typeclass.ml:
```diff
@@ -259,54 +259,56 @@ let make_method loc cl_num expr =
+           let (cty, cty') = type_constraint env sty sty' ctf.pctf_loc in
+           mkctf (Tctf_constraint (cty, cty')))
 
   | Pctf_attribute x ->
```
It took me a few seconds to realize that it's ok to ignore pctf_attributes on the Pctf_attribute node, even though they could in theory affect warnings triggered by the processing of the Pctf_attribute payload.  The reason is that the concrete syntax does not allow attributes on Pctf_attribute.  A PPX could insert some, but I think it's ok to ignore them.

alainfrisch commented:
It could be worth adding a test to show the effect of d3763f1d29d00d7c7694870c097823954f310b9d (likely some code that becomes ill-typed).

lpw25 commented:
Oh yes, I missed that one. I'll add some tests.

alainfrisch commented:
Just to clarify: it seems that csig_inher is unused currently (not a consequence of other changes), or did I miss something?

alainfrisch commented:
I skimmed through each commit, and they all look very reasonable to me, although I don't have the expertise to review them in depth.  Anyway, thanks @lpw25 for splitting the large contribution into individual commits, with good explanations and tests to show the user-side of the changes.

alainfrisch commented:
Some explanation about the dummy method could be useful.  Is it a hack to benefit from level (now scope) analysis, and could it be replaced by something more explicit?

lpw25 commented:
> Just to clarify: it seems that csig_inher is unused currently (not a consequence of other changes), or did I miss something?

Yeah, it is already unused

alainfrisch commented on typing/typedtree.mli:
```diff
@@ -215,7 +215,7 @@ and expression_desc =
       Ident.t * Parsetree.pattern * expression * expression * direction_flag *
         expression
-  | Texp_send of expression * meth * expression option
+  | Texp_send of expression * meth
```
I've wondered several times what was this `expression option` (and always had to go back to the code to figure out).  Very nice to see it going away.

gasche replied:
Seconded, I also tried (not very successfully) to figure this out as part of the "recursive value declarations" work.

lpw25 commented:
> Some explanation about the dummy method could be useful. Is it a hack to benefit from level (now scope) analysis, and could it be replaced by something more explicit?

It has two purposes: it stops the self types from becoming closed and it stops the self types from escaping the scope of the class declaration. It could be replaced by something more explicit, although the fundamental idea: that the restriction on closing be represented by some form of field of the object type is probably the right way to do things.

alainfrisch commented:
Not directly related to your commits, but I've always found it weird that cl_num is represented as a string (which is always a stringified  int, right?). What about introducing at least a type alias for the cl_num fields in Types (Val_self/Val_anc), perhaps switching to something else than string?   And/or switching those constructors to use inline records, to document the meaning of their arguments.

lpw25 commented:
I'd like to tidy that up at some point, but I'd rather not do it in this PR -- which already rewrites a lot of the classes code.

lpw25 committed:
Add helpers for warning_scope in typeclass.ml

Add a couple of small helper functions to reduce code duplication
around calls to warning_scope.

alainfrisch commented:
> It has two purposes

Thanks!  It could be useful to write this down in the code, if you find a natural place to put it.

> I'd rather not do it in this PR 

Fair enough.

lpw25 commented:
I've added tests for the change to instance variables in class types.

garrigue commented:
I have looked a bit at the commits to find what you are trying to fix, but this didn't help me much.
You add many tests, but there is no clear statement of what you expect to be the correct outcome of the test. The class system being underspecified, we might not agree on what it should be :)
Same problem with warnings. For instance, you choose to warn as soon as unification introduces a non-declared virtual method. The previous behavior of only warning when they were introduced by a method call was intentional, to avoid typos. The goal is less clear in you example.
I think we need a list of all changes, and why they were introduced, so that we can discuss them properly.
As always, I would have preferred discussing the specification issues before writing the code, rather than now with this huge PR.

From a different perspective, my impression was that the class language was stable, after many years of fixing bugs, and the best thing to do was to leave it alone as long as we don't want to introduce new features. You are clearly taking a wholly different approach, but does it mean that you want to introduce new features?

lpw25 commented:
> I think we need a list of all changes, and why they were introduced, so that we can discuss them properly.

I think that the list of specification changes -- i.e. things that are not just obvious bug fixes -- is as follows:

#### Extend warning 17 to catch indirect cases
Currently warning 17 occurs when attempting to directly send a method to self that has not been declared:
```ocaml
# class virtual c = object (self) initializer self#foo end;;
Characters 44-52:
  class virtual c = object (self) initializer self#foo end;;
                                              ^^^^^^^^
Warning 17: the virtual method foo is not declared.
class virtual c : object method private virtual foo : unit end
```
However, it does not detect the same thing happening indirectly:
```ocaml
# let send_foo o = o#foo;;
val send_foo : < foo : 'a; .. > -> 'a = <fun>
# class virtual c = object (self) initializer send_foo self end;;
class virtual c : object method virtual foo : unit end
```
This PR addresses that short-coming:
```ocaml
# let send_foo o = o#foo;;
val send_foo : < foo : 'a; .. > -> 'a = <fun>
#   class virtual c = object (self) initializer send_foo self end;;
Line 2, characters 18-61:
2 | class virtual c = object (self) initializer send_foo self end;;
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Warning 17: the virtual method foo is not declared.
class virtual c : object method virtual foo : unit end
```

#### Check for warning 15 in more cases
Warning 15 detects when a private method is implicitly made public:
```ocaml
# class c = object(self) method private foo = () initializer send_foo self end;;
Characters 10-76:
  class c = object(self) method private foo = () initializer send_foo self end;;
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Warning 15: the following private methods were made public implicitly:
 foo.
class c : object method foo : unit end
```
However, it currently only applies for class structures, failing to detect cases like:
```ocaml
# class ['a] c = object (_ : 'a) method private foo = () end;;
class ['a] c :
  object ('a) constraint 'a = < .. > method private foo : unit end
# class d = [ < foo : unit; .. > ] c;;
class d : object method foo : unit end
```
and it is applied for class types even though the same issue occurs:
```ocaml
# class type c = object (< foo : unit; ..>) method private foo : unit end;;
class type c = object method foo : unit end
```
This PR addresses those short-comings:
```ocaml
# class d = [ < foo : unit; .. > ] c;;
Line 1, characters 10-34:
1 | class d = [ < foo : unit; .. > ] c;;
              ^^^^^^^^^^^^^^^^^^^^^^^^
Warning 15: the following private methods were made public implicitly:
 foo.
class d : object method foo : unit end
# class type c = object (< foo : unit; ..>) method private foo : unit end;;

Line 1, characters 15-71:
1 | class type c = object (< foo : unit; ..>) method private foo : unit end;;
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Warning 15: the following private methods were made public implicitly:
 foo.
class type c = object method foo : unit end
```

#### Relax object duplication restrictions

Currently object duplications (i.e. `{< ... >}`) can only refer toinstance variables that are specified before the method in which they occur:
```ocaml
# class c = object method foo = {< x = 5 >} val x = 0 end;;
Characters 30-41:
  class c = object method foo = {< x = 5 >} val x = 0 end;;
                                ^^^^^^^^^^^
Error: Unbound instance variable x
```
This restriction made sense before OCaml 3.10 when instance variables could shadow one another. However, instancevariables now override one another instead and so the restriction is not needed:
```ocaml
# class c = object method foo = {< x = 5 >} val x = 0 end;;
class c : object ('a) val x : int method foo : 'a end
```

#### Override instance variables in class types
Currently, the behaviour of repeated instance variable definitions in class types is very odd: it partially shadows and partially overrides the previous definition. The best example of this behaviour is probably:
```ocaml
# class type c = object val x : int val virtual x : float end;;
class type c = object val x : float end
```
This is probably a hangover from the change of behaviour in 3.10. Fixing this mixed behaviour is probably a bug-fix, however I've chosen to fix it by making repeated definitions override previous definitions -- which is less flexible than shadowing but more consistent with the behvaviour is classes -- and that's probably a change in specification.

#### Remove dummy methods from classes earlier

I've moved the removal of dummy methods from classes to earlier in the process, which allows more definitions such as:
```ocaml
class c = object method m = 5 end
  and d = object method n = object inherit c end end;;
```
that mix mutual class recursion and object literals.

lpw25 commented:
> As always, I would have preferred discussing the specification issues before writing the code, rather than now with this huge PR.

I did not set out to make changes to the specification, I just wanted to make the code clearer and fix one bug. But, as always with the compiler, once I understood the code it was clear there were a number of bugs and undesirable behaviours. All the changes to the specification above are obvious once the code is made clearer -- it becomes clear that we are checking something in unnatural place or adding ad-hoc difference between equivalent operations -- and so by making them you improve the specification and further improve the code.

garrigue commented:
Thank you for this list. This clarifies a lot.

Concerning warning 17, this may be open to debate, but I believe we should stick to the original behavior: there is nothing in the language that says that virtual methods should only be declared with the `virtual` keyword; this warning is rather about protecting against typos in calls to self-methods in the body of a class. So maybe the name of the warning is wrong, but I don't think its extent should change.

Concerning warning 15, again this is open to debate. I would be more inclined to side with you on that one, but we must see the implications.

Concerning the relaxation of object duplication, I'm not sure. The restriction is not completely unreasonable (usually one declares instance variables first, so this shouldn't happen anyway), but I agree this would be fine to remove it. Removing it of course has a cost, which disappears if you have two passes.

Concerning the overriding of instance variable declarations, I couldn't understand what is the new behavior: what is the difference between shadowing and overriding here? I agree this is a left-over of the previous behavior, but not dangerous. Note also that through a class type coercion one can hide an instance variable, and create later a new one with the same name, so changing the type of an instance variable is not meaningless, and allowing it in class types may indeed be a comfort, so changing this behavior could be problematic.

Last, for  your examples allowed by removing the dummy method earlier, they seem indeed nice.
We could live with the limitation, but removing it is an improvement.

First, we need to decide on all these issues. Then I suppose that the improvements provided could make these changes worthwhile, despite the risks. The point being that improving maintainability is great, but if there is no maintenance needed to start with it may be both gratuitous and dangerous.

I'm still quite afraid of the size of your changes. I can try to read the code, but I'm not sure I could make a proper review. Fortunately you added lots of tests, which is a major plus. However the output changes on a lot of existing tests, which makes the checks difficult. And we also need to verify that this doesn't break any existing code in the wild.

I'm also concerned that your patch clearly changes the sharing inside types, as the results of tests show. This may be related to the redundancy you introduce in the representation of class signatures. I'm not sure how to investigate that, and I do not see a clear impact other than the printed types, so maybe it's fine. (I'm assuming that all tests run both in standard an principal mode, which should be the case with expect tests.)

lpw25 commented:
> Concerning warning 17...

I don't know what the original motivation for the warning was, but I disagree with you about what the purpose of the warning should be. The ability to implicitly declare virtual methods is, in my opinion, a confusing misfeature of the language. This warning tells you when this misfeature has occurred allowing you to avoid it, or to at least understand when it is the source of a confusing error. Besides, typos are no less likely to occur indirectly than directly, so even with your proposed meaning for this warning the change is an improvement.

> Removing it of course has a cost, which disappears if you have two passes.

It's probably worth pointing out that we already had two passes really, its just that they were written using laziness rather than as explicit passes. This PR makes all this explicit.

> Concerning the overriding of instance variable declarations, I couldn't understand what is the new behavior: what is the difference between shadowing and overriding here? 

So the current behaviour is that the type and mutability are taken from the second definition, whilst the "virtual-ness" is the least-upper bound from both definitions. The new behaviour is that the types of both definitions are unified, the mutability of both definitions must agree and the "virtual-ness" is the least-upper bound from both definitions.

The new behaviour makes sense -- it matches the behaviour in classes. Another reasonable behaviour would be to take everything from the second definition and ignore the first. As far as I can tell, the current behaviour makes no sense and is probably a bug.

> The point being that improving maintainability is great, but if there is no maintenance needed to start with it may be both gratuitous and dangerous.

There is always maintenance needed. When you leave bad code around in an evolving code base it always ends requiring changes, and since the code is hard to understand people try to make the smallest changes possible (i.e. hacks), so the code always just gets worse over time.

The changes in 4.07 are a great example. In order to get disambiguation of GADTs without breaking people's code we needed to make a slight change to how object literals were handled. Because the code was hard to understand we ended up putting in a number of bugs. If the code had been easy to understand that wouldn't have happened, and if we leave typeclass.ml in the state it is in then it will certainly happen again. I can't imagine managing to add modular implicits, algebraic effects, etc. without needing to make at least some changes to this file.

It also prevents making even small improvements to how objects and classes are handled. For example, see #7159, where you say:

> I see your point.
> As you point out, it shouldn't be that difficult, since one should just propagate the expected type to the internal self-type.
> However, code in typeclass is really hairy, so don't hold your breath.

with this PR that change would be relatively simple (I considered including it in this PR but decided there was already enough in here).

> I'm also concerned that your patch clearly changes the sharing inside types, as the results of tests show. 

I don't think it changes much about the sharing inside the types. It's more that it adds another copy of the method types, and that copy is used for printing which obviously affects the output. The change in
[this commit](https://github.com/ocaml/ocaml/pull/8516/commits/d8cab4df0c2746f9b6c3f85072948c7c76830342) also affects the printing.

> I'm still quite afraid of the size of your changes. I can try to read the code, but I'm not sure I could make a proper review.

I would suggest just going one commit at a time. I think the only difficult commits are going to be:

- [Change representation of class types](https://github.com/ocaml/ocaml/pull/8516/commits/3ce3f2033b983d6908a974155ebecd0dad549bce)
- [Treat class_signature more like type_expr](https://github.com/ocaml/ocaml/pull/8516/commits/2e794bf52fbdabe1dde4acd35d9ccf7bd5e925fd)

the rest are fairly small and self-contained. We can always split parts off into other PRs and merge them as we go. We can also undo changes to the specification if needed and then put them in a separate PR.

(Note that I just did a rebase and force push to fix the order that the commits are displayed by github -- they should now appear in the correct order.)

garrigue commented on typing/typeclass.ml:
```diff
@@ -1566,9 +1578,16 @@ and class_expr_aux cl_num val_env met_env scl =
       rc {cl_desc = Tcl_constraint (cl, Some clty, vals, meths, concrs);
           cl_loc = scl.pcl_loc;
-          cl_type = snd (Ctype.instance_class [] clty.cltyp_type);
+          cl_type = ty;
```
How is this line related with the addition of the dummy method?

garrigue commented on testsuite/tests/typing-objects/Tests.ml:
```diff
@@ -988,11 +988,8 @@ class type ['a] ct = object ('a) constraint 'a = < .. > end
-Error: This class should be virtual. The following methods are undefined : a
+Error: This class has undeclared virtual methods.
+       The following virtual methods were not declared : a
 |}];;
```
This error message is strange: there is a warning for undeclared virtual methods, but this is not an error.

garrigue commented on testsuite/tests/typing-gadts/pr7260.ml:
```diff
@@ -25,14 +25,6 @@ Line 8, characters 2-96:
-       The following methods are undefined : bar
+Error: This class has undeclared virtual methods.
+       The following virtual methods were not declared : bar
 |}];;
```
This error message is strange: there is a warning for undeclared virtual methods, but this is not an error.

garrigue commented on typing/ctype.ml:
```diff
@@ -835,7 +836,7 @@ let rec update_level env level expand ty =
     | Tfield(lab, _, ty1, _)
-      when lab = dummy_method && (repr ty1).level > level ->
+      when lab = dummy_method && level < (repr ty1).scope ->
         raise Trace.(Unify [escape Self])
```
Good idea. The prohibition to lower the level of the dummy method (even locally) was a real pain.

garrigue commented on typing/typemod.ml:
```diff
@@ -1543,35 +1542,35 @@ let path_of_module mexp =
 
 (* Check that all core type schemes in a structure are closed *)
 
```
Should probably change this comment too.

garrigue commented on utils/warnings.ml:
```diff
@@ -441,13 +441,11 @@ let message = function
   | Instance_variable_override [lab] ->
-      "the instance variable " ^ lab ^ " is overridden.\n" ^
-      "The behaviour changed in ocaml 3.10 (previous behaviour was hiding.)"
+      "the instance variable " ^ lab ^ " is overridden."
```
The fact 3.10 was a long time ago doesn't mean there is no old code left around.
(Yet there was no warning for the previous transition, from OCaml 1 to OCaml 2.00, when the behavior changed from overriding to shadowing.)

garrigue commented on ocamldoc/odoc_print.ml:
```diff
@@ -95,9 +95,8 @@ let simpl_class_type t =
-            csig_inher = []; }
+            csig_vars = Types.Vars.empty;
+            csig_meths = Types.Meths.empty; }
         in
```
`csig_inher` was intended to provide information about the inheritance hierarchy of a class type.
This is certainly useful information, and actually I was convinced that ocamldoc used it.
I believe Maxence asked me to add it.
Before removing it, it might be worth thinking again whether somebody has a use for it.

garrigue commented:
> > Concerning warning 17...
> 
> I don't know what the original motivation for the warning was, but I disagree with you about what the purpose of the warning should be. The ability to implicitly declare virtual methods is, in my opinion, a confusing misfeature of the language. This warning tells you when this misfeature has occurred allowing you to avoid it, or to at least understand when it is the source of a confusing error. Besides, typos are no less likely to occur indirectly than directly, so even with your proposed meaning for this warning the change is an improvement.

Basically, your change means that all virtual methods must be explicitly declared.
In particular, you would have a warning with the following code:
```
type t = < a : int >
class virtual c = object (_ : <t; ..>) end
```
This code uses "interface inheritance through self", and I believe it is completely legit.

Of course one can make typos anywhere. But usually one is more careful when writing types than method calls, so I don't see the two problems are equivalent.

At the very least I would suggest moving this change to a separate PR, where we can discuss it independently. Same thing for warning 15, since this is basically the same problem (manipulating the interface through direct constraints to self).

> > Removing it of course has a cost, which disappears if you have two passes.
> 
> It's probably worth pointing out that we already had two passes really, its just that they were written using laziness rather than as explicit passes. This PR makes all this explicit.

Of course. Actually there were two passes in ocaml 1.00, and Jérôme changed that to one pass in ocaml 2.00, which resulted in using `lazy` when introducing polymorphic methods.
I should probably have reverted to two passes at that point, but somehow the changes where very circumscribed.

> > Concerning the overriding of instance variable declarations, I couldn't understand what is the new behavior: what is the difference between shadowing and overriding here?
> 
> So the current behaviour is that the type and mutability are taken from the second definition, whilst the "virtual-ness" is the least-upper bound from both definitions. The new behaviour is that the types of both definitions are unified, the mutability of both definitions must agree and the "virtual-ness" is the least-upper bound from both definitions.
> 
> The new behaviour makes sense -- it matches the behaviour in classes. Another reasonable behaviour would be to take everything from the second definition and ignore the first. As far as I can tell, the current behavior makes no sense and is probably a bug.

OK. I'm fine with the new behavior, but I would still suggest moving it to a separate PR, as there is a potential for breakage here.

> > The point being that improving maintainability is great, but if there is no maintenance needed to start with it may be both gratuitous and dangerous.
> 
> There is always maintenance needed. When you leave bad code around in an evolving code base it always ends requiring changes, and since the code is hard to understand people try to make the smallest changes possible (i.e. hacks), so the code always just gets worse over time.

I kind of agree. But I'm not sure your extensive changes remove _all_ the mystery in typeclass.
Thank you for the courage anyway, I'm just too much of a coward.

> > I'm also concerned that your patch clearly changes the sharing inside types, as the results of tests show.
> 
> I don't think it changes much about the sharing inside the types. It's more that it adds another copy of the method types, and that copy is used for printing which obviously affects the output. The change in
> (this commit)[https://github.com/[/pull/8516](https://github.com/ocaml/ocaml/pull/8516)/commits/d8cab4df0c2746f9b6c3f85072948c7c76830342] also affects the printing.

Did you measure the change in the size of cmi's, for lablgtk for instance?
This was a real concern at some point.
Also, there is some logic in `Subst` to avoid duplicating too much (see "Do not clean up if saving"), but it may have suffered from bit-rot, and might be better removed if it doesn't help anymore.

> > I'm still quite afraid of the size of your changes. I can try to read the code, but I'm not sure I could make a proper review.
> 
> I would suggest just going one commit at a time. I think the only difficult commits are going to be:

I did go through each commit, adding a few comments.
My overall impression is that what you are doing is reasonable and helpful, but I think it is just too big for me to do a proper review (sorry, I'm a bad code reviewer).
If somebody is ready to go through all the code with good eyes, this might help (@trefis ?).
Discussing it directly might also work.

At this point, my only request would be to move the above changes of behavior to separate PRs, so that we can first fix the real bug this PR was supposed to fix.

lpw25 commented:
Thanks for the preliminary review. I'm happy to move the more controversial parts into their own PRs. I'm on holiday for a few weeks, so it will be a little while before I get around to it.

trefis committed:
Treat class_signature more like type_expr

Treat class_signature more like type_expr by making all its
components mutable and using three unification-like operations
for manipulating them: add_method, add_instance_variable and
inherit_class_signature. These operations behave similarly to
filter_self_method, which they replace.

We move much of the logic for handling class_signature into Btype
and Ctype and use it for the typing of class signatures
and class structures.

Instead of using method and variable tables with both identifiers and
type components as accumulators during class structure typing, we use
tables with just identifiers along with a single class_signature. This
makes the logic clearer and makes it easier to share things with the
typing of class signatures.

lpw25 committed:
Build met_env on second pass

Move the extension of the method environment to the second pass of
`class_field` in "typeclass.ml".

lpw25 committed:
Add Changes entry

lpw25 commented:
I've rebased this PR -- based on an initial rebase by @trefis. I've also addressed comments and removed the changes to the behaviour of warnings 15 and 17. @trefis has also promised me a review, so I've added a Changes entry attributing the review to him.

garrigue commented on typing/ctype.ml:
```diff
@@ -2723,9 +2680,7 @@ and unify3 env t1 t1' t2 t2' =
-        if is_self_type d1 (* PR#7711: do not abbreviate self type *)
-        then link_type t1' t2'
-        else link_type t1' t2
+        link_type t1' t2
```
Does it mean that you solved the problem in #7711 ?

lpw25 replied:
Yes. There are now no special syntactic restrictions on self types.

garrigue reviewed (commented):
Still trying where to start reading...

trefis commented on typing/typeclass.ml:
```diff
@@ -570,237 +413,548 @@ and class_type_aux env scty =
+let enter_instance_var_val name val_env =
+  Env.enter_unbound_value name Val_unbound_instance_variable val_env
+
+let enter_ancestor_met loc name sign meths cl_num ty attrs met_env =
```
Nitpick: if you turned the `attrs` parameter into a labelled one, callers which give a default value (`[]`) would be slightly more readable.

lpw25 replied:
Done

trefis commented on testsuite/tests/typing-misc/includeclass_errors.ml:
```diff
@@ -146,7 +146,8 @@ Lines 2-5, characters 4-7:
 5 |     end
-Error: The class type object method foo : string end
+Error: The class type
+         object method private virtual cast : int method foo : string end
```
This is not your fault, but this diff would have been easier to read if `foo_t` had been defined just above `foo` (the only place where it's used) instead of at the top of the file.

lpw25 replied:
`foo_t` seems to be used twice so I just left it where it is.

trefis commented on typing/ctype.ml:
```diff
@@ -667,34 +657,22 @@ let closed_extension_constructor ext =
+    Meths.iter
+      (fun lab (priv, _, ty) ->
+        if priv = Public then
         try closed_type ty with Non_closed (ty0, real) ->
```
While you're there, can you fix the indentation?

lpw25 replied:
Done

trefis commented on typing/ctype.ml:
```diff
@@ -3772,125 +3770,112 @@ let rec moregen_clty trace type_pairs env cty1 cty2 =
-        (fun (lab, k, _) err ->
+  let sign1 = signature_of_class_type pat_sch in
+  let sign2 = signature_of_class_type subj_sch in
+  let error = [] in
```
Two bikeshedding comments here:
1. I'd like "errors" better than "error", given that we're threading a list of errors. (Yes, the name was already there, but ...)
2. I'm not fond of the `let error = []`. I understand the regularity with the code that follows, but meh.

:slightly_smiling_face:

lpw25 replied:
Done

trefis commented on typing/ctype.ml:
```diff
@@ -3907,19 +3892,13 @@ let rec equal_clty trace type_pairs subst env cty1 cty2 =
-          pairs;
+        Meths.iter
+          (fun lab (_, _, ty) ->
+             let (_, _, ty') = Meths.find lab sign1.csig_meths in
```
I'd add a comment here saying that this call to `Meths.find` won't raise because the only caller of this function (`match_class_declarations`) errors in case of missing methods.

(The previous version of the code would also have deserved a comment)

lpw25 replied:
Done

trefis commented on typing/ctype.ml:
```diff
@@ -3936,62 +3915,48 @@ let rec equal_clty trace type_pairs subst env cty1 cty2 =
+           | Virtual -> CM_Hide_virtual ("method", lab) :: err
+           | Concrete -> err
+         end)
+      sign1.csig_meths error
```
This block (and some others in this fonction) looks very similar to one already present in `match_cltype`.
Could these auxiliary checks be shared?

lpw25 replied:
I lifted all the shared stuff out into a separate function.

trefis commented on typing/typeclass.ml:
```diff
@@ -2271,23 +2322,35 @@ let report_error env ppf = function
+        match kind with
+        | Object -> "object"
+        | Class -> "non-virtual class"
+        | Class_type -> "non-virtual class type"
```
Duplicated with the case above.
Add a `string_of_kind` or `pp_kind` helper perhaps?

lpw25 replied:
Done

trefis commented on typing/typeclass.ml:
```diff
@@ -1473,8 +1460,12 @@ let initial_env define_class approx
-  if !Clflags.principal then Ctype.generalize_spine constr_type;
+  if !Clflags.principal then begin
+    Ctype.end_def ();
+    Ctype.generalize_structure constr_type;
```
Cf. https://github.com/ocaml/ocaml/issues/5545#issuecomment-472957253 for a confirmation that this is correct.

trefis commented on typing/ctype.ml:
```diff
@@ -1033,6 +973,21 @@ let limited_generalize ty0 ty =
+        sign.csig_meths
+  | Cty_arrow (_, ty, cty) ->
+      limited_generalize rv ty;
+      limited_generalize_class_type rv cty
```
Isn't this `generalize_class_type' (limited_generalize rv)`?

lpw25 replied:
Well spotted. That's definitely simpler.

trefis commented on typing/ctype.ml:
```diff
@@ -3259,24 +3207,266 @@ let filter_method env name priv ty =
+let update_class_signature env sign =
+  let self = expand_head env sign.Types.csig_self in
+  let fields, row = flatten_fields (object_fields self) in
+  let meths, implicit_public, implicit_declared =
```
Nitpick: `implicitly_declared`*?
(I'd also suggest `implicitly_public` but then the error message talks about "implicit public methods" so it's less obvious :))

lpw25 replied:
Done

trefis commented on typing/ctype.ml:
```diff
@@ -4520,47 +4678,71 @@ let cyclic_abbrev env id ty =
+  | Cty_constr (_, params, _) ->
+      List.exists (nongen_schema Env.empty) params
+  | Cty_signature sign ->
+      nongen_schema Env.empty sign.csig_self
+      || nongen_schema Env.empty sign.csig_self_row
+      || Meths.exists
+           (fun _ (_, _, ty) -> nongen_schema Env.empty ty)
+           sign.csig_meths
```
IIUC there should be some redundancy between checking the self type / row type and each of the methods.
I'm not sure I have a clear picture of when these types aren't simply subsumed by the information contained in `csig_meths`. There are cases where the self type is abbreviated (in which case there shouldn't be any duplicated work), anything else?

lpw25 replied:
The check on the row type should always be redundant since it is within the self type, but I prefer to include it and have the code obviously correct rather than subtly correct. The check on self is redundant when we have called `update_class_signature`, which we will always have done when we call this function, but again I'd rather leave the redundancy and not rely on that.

trefis commented on typing/typeclass.ml:
```diff
@@ -570,237 +413,548 @@ and class_type_aux env scty =
+           end;
+           if !Clflags.principal then Ctype.begin_def ();
+           let definition = type_exp val_env sdefinition in
+           if !Clflags.principal then begin
+             Ctype.end_def ();
+             Ctype.generalize_structure definition.exp_type
+           end;
+           add_instance_variable loc val_env
+             label.txt mut Concrete definition.exp_type sign;
+           let already_declared, val_env, par_env, id, vars =
+             match Vars.find label.txt vars with
+             | id -> true, val_env, par_env, id, vars
+             | exception Not_found ->
+                 let name = label.txt in
+                 let val_env = enter_instance_var_val name val_env in
+                 let par_env = enter_instance_var_val name val_env in
+                 let id = Ident.create_local name in
+                 let vars = Vars.add label.txt id vars in
+                 false, val_env, par_env, id, vars
```
Duplicated from the virtual case.

lpw25 replied:
I couldn't find a way to share the lines that I liked, so I've left them duplicated.

trefis reviewed (approved):
I left a couple of comments which, worryingly for a PR of this size, are all "minor" comments. I'm not sure what I should deduce from this. That @lpw25 produces very good code? That I missed some subtle (or obvious?) issues? Both? Neither?

I read this PR commit by commit, which I think was the only reasonable way to review it, but also introduced some extra work (I left a bunch of comments and questions on code that later disappeared!).
Most of the commits are easy to review, the notable exceptions are:
- 8101d5279b7e59637f0988bd8fc3ed2242db51c2 "Change representation of class types"
- 035eff2ede825363a9d14e02d77115162d0f1003 "Treat class_signature more like type_expr"

which moved a lot of code around while also changing some of it. If someone else (@garrigue?) wanted to give a second look  at this PR, I'd suggest prioritizing those.

My general feeling towards this PR is that while I find the style of the new code a bit heavy, it's definitely easier to follow what's going on. (So I refrained from leaving style comments)

trefis commented:
FTR some good test cases to look at before merging:
- lablgtk (ping @garrigue ?)
- uses of [visitors](https://gitlab.inria.fr/fpottier/visitors). @fpottier would building the testsuite of the visitors package be enough to get a good coverage, or do you know how specific users which would be more interesting?

fpottier commented:
>   Would building the testsuite of the visitors package be enough to get a good coverage, or do you know how specific users which would be more interesting?

Building the testsuite would be a good start. If you want to go further, heavy users of `visitors` include Frama-C and Rotor, I think.

lpw25 commented:
Rebased and review comments addressed.

trefis commented:
I couldn't test this on visitors, or rather, I couldn't run visitors testsuite because it depends on too many things which aren't available on 4.13 (ppx_import, core_bench).
But I did try and build lablgtk with this PR, both master and the latest released version and got the same error on both:
```
File "gBin.mli", line 186, characters 4-34:
186 |     val obj : Gtk.aspect_frame obj
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: The instance variable obj has type
         Gtk.aspect_frame Gtk.obj = Gtk.aspect_frame Gobject.obj
       but is expected to have type Gtk.frame Gtk.obj = Gtk.frame Gobject.obj
       Type
         Gtk.aspect_frame =
           [ `aspectframe | `bin | `container | `frame | `gtk | `widget ]
       is not compatible with type
         Gtk.frame = [ `bin | `container | `frame | `gtk | `widget ] 
       The second variant type does not allow tag(s) `aspectframe
```
which is a consequence of 95621fe.
And can be fixed with:
```
diff --git a/src/gBin.mli b/src/gBin.mli
index 9eeb4631..f2ea0b82 100644
--- a/src/gBin.mli
+++ b/src/gBin.mli
@@ -182,8 +182,9 @@ val frame :
    @gtkdoc gtk GtkAspectFrame *)
 class aspect_frame : Gtk.aspect_frame obj ->
   object
-    inherit frame
+    inherit frame_skel
     val obj : Gtk.aspect_frame obj
+    method connect : GContainer.container_signals
     method set_obey_child : bool -> unit
     method set_ratio : float -> unit
     method set_xalign : float -> unit
```
(though perhaps there is a more elegant fix).
Apart from that, everything built fine.

I find this (lone) error acceptable and am tempted to merge the PR.
@garrigue: any objection?

---

FTR, I've reduced the example to this:
```ocaml
type foobar = [ `foo | `bar ]
type foobarbaz = [ foobar | `baz ]

module type test = sig 
  class c0 : 'a list -> object
    constraint 'a = [> foobar ]
    val x : 'a list
  end

  class c1 : foobar list -> object
    inherit c0
    val x : foobar list
  end

  class c2 : foobarbaz list -> object
    inherit c1
    val x : foobarbaz list
  end
end
```

garrigue commented:
@trefis Thanks for checking.
As always breaking changes are a pain, but I think I already accepted the logic behind this one. So if the only thing broken is lablgtk, this is fine.
Still, this brings the question of whether we should test this on opam, and if so, how, due to compatibility problems. One way could be to wait after after 4.13 is released to be able to check more software, but 4.13 and trunk have already started to diverge.

trefis commented:
Update:

> I couldn't test this on visitors, or rather, I couldn't run visitors testsuite because it depends on too many things which aren't available on 4.13 (ppx_import, core_bench).

With some guidance from @Octachron I managed to get visitors testsuite to run (modulo the ppx_import test), and it ran fine.

trefis commented:
Here is, I believe, a fair summary of the state of this PR:
- rebasing it is costly; it's been rebased recently; it already has some minor conflicts with trunk (but applies cleanly on 4.13)
- it's been reviewed
- it adds a bunch of new tests
- it's been tested on two big public users of classes/objects, and no bugs where found
- it implies a few observable changes, which are listed/described in [this comment](https://github.com/ocaml/ocaml/pull/8516#issuecomment-474386718), one of them is a breaking change.
  This breaking change:
  + did break one of the packages tested
  + is only a couple of lines; easily reverted independently of the rest of the PR if we ever wished to do so

As a result, I would be in favor of merging as soon as this PR is rebased (once more).
The breaking change can always be reconsidered at a later stage if it proves to break too many opam packages or some other private repositories (ping @alainfrisch? This PR applies cleanly on 4.13 right now, if you wish to test it).

gasche commented:
A comment from a "co-decider in a hurry" (I have forgotten about the details of this PR):
- You point to a comment with 4 changes and you mention that one of them is breaking, but I'm not sure which and it's not obvious. Quoting this change in full would help have a self-contained summary.
- If it's easy to not include the breaking change in the PR, what about... not including it for now?
- I get the "rebasing is painful" argument, if it has been reviewed we should merge quickly. Merging the non-breaking part right away is obviously a good decision. No strong opinion on the breaking stange.
- For the breaking change, if I understand correctly it wouldn't be so hard to do opam-wide testing (or would it?), and/or it may be easier to get data once the opam ecosystem gets more compatible with 4.13. Could we get opam feedback before merging it?

(Recently our track record of dealing with compatibility issues has been weird, in the sense that we have tended to introduce problems by being overly cautious about backward-compatibility (#9811 and related issues). I would still be in favor of following due process, if only because it's a good habit to maintain.)

trefis commented:
> - You point to a comment with 4 changes and you mention that one of them is breaking, but I'm not sure which and it's not obvious. Quoting this change in full would help have a self-contained summary.

I thought it was :cold_sweat: but fair point.

> - If it's easy to not include the breaking change in the PR, what about... not including it for now?

Actually, @garrigue already suggested it should be split into its own PR.
@lpw25 can you just do that?

> - I get the "rebasing is painful" argument, if it has been reviewed we should merge quickly. Merging the non-breaking part right away is obviously a good decision. No strong opinion on the breaking stange.

:+1:

> - For the breaking change, if I understand correctly it wouldn't be so hard to do opam-wide testing (or would it?), and/or it may be easier to get data once the opam ecosystem gets more compatible with 4.13. Could we get opam feedback before merging it?

opam-wide testing is probably a bit difficult right now, I don't know which part of the universe is 4.13 compatible already.
But let's move that discussion to a dedicated PR when it appears.

lpw25 committed:
Change representation of class types

Previously, class types represented their methods via the csig_self
field. This was a type_expr that was restricted to be syntactically
a Tobject node. With this patch the methods are represented directly
with a methods table. csig_self is no longer restricted to be
a Tobject node and is no longer required to contain the private methods.
We also add a csig_self_row field to hold the row variable of the class
type -- which means we can avoid going through csig_self to find it.

lpw25 committed:
Bootstrap

lpw25 commented:
Ok, I've rebased and I've removed the change that was breaking lablgtk

gasche commented:
The bug/regression #10781, new in 4.14+dev, may be related to the present PR.
